Reworking critical technology concepts (#217)
DHBW-FN/OxideWM: concepts/ipc.md

# IPC

> During a first discussion it was decided that oxide should be controllable via an ipc mechanism.

Reviewer: ```suggestion
During a first discussion it was decided that *Oxide* should be controllable via an ipc mechanism.
```

* External libraries
* Command line utility

> ## Feature List

Reviewer: ```suggestion
## Feature list
```

The following list includes currently proposed features

> * Everything that is achievable via the keyboard (kill, move, launch...)

Reviewer: ```suggestion
* everything that is achievable via the keyboard (kill, move, launch...)
```

> * Current state of the window manager including layout, windows...

Reviewer: ```suggestion
* current state of the window manager including e.g. layout or windows
```

> ## IPC Integration Solution

Reviewer: ```suggestion
## IPC integration solution
```

> Since there are two types of events that have to be handeled, there

Reviewer: ```suggestion
Since there are two types of events that have to be handled, there
```

> needs to be some separation bewteen them.

Reviewer: ```suggestion
needs to be some separation between them.
```

One type are `xevents`, received from the `X11` instance, and the other one

> is custom events create by the user, recieved over `zbus`.

Reviewer: ```suggestion
is custom events created by the user, received over `zbus`.
```

For this reason, each type of event will get its own loop on its own thread,
which will await them and push them into a list shared between them.
The events in this list will be taken care of by the window manager,
who will execute the correct action based on event type and content.

![IPC-QUEUE](ipc-queue.png)

> ## Technical Solution

Reviewer: ```suggestion
## Technical solution
```

The following sections describe the argument for the different ipc-mechanisms and libraries.

### Requirements
As for the aforementioned use cases it will not be required to send large amounts of data. 
Only short messages will be exchanged between the clients. Also it is not expected that the ipc performance will have a significant impact on the usability of the system.
Therefore some ipc options such as shared memory and semaphores will not be regarded as these options are not as easy to use and do not offer any significant advantages.

> ### Possible IPC Mechanisms

Reviewer: ```suggestion
### Possible IPC mechanisms
```

> * D-Bus should offer a high level, easy to use ipc mechanism

Reviewer: The word "easy" must not be contained in any of our documents. Please think about another formulation :)

Reviewer 2: "simple"

### Conclusion

After a technical discussion with our team, we came to the conclusion that **D-Bus is most suitable**.
The **performance is deemed non critical** in our use case and the ease of use will be benefitial for the project.
None the less, the ipc interface should be **created in an abstract manner** allowing for a possible replacement of the underlying ipc mechanism.  

### Implementation

> #### Available Libraries

Reviewer: ```suggestion
#### Available libraries
```

* Wrapper library for libdbus -> libdbus dependency
* Examples

#### Conclusion
Zbus seems to have some advantages over dbus-rs, mainly:

> * Official freedesktop.org library

Reviewer: ```suggestion
* official freedesktop.org library
```

> * Pure rust -> no ibdbus dependency

Reviewer: ```suggestion
* pure rust -> no ibdbus dependency
```

> * Pure rust -> no ibdbus dependency

Reviewer: Is the term "ibdbus" correct? I don't know it and i am not sure if it is a typo or not
